rfc: Add P2P info exchange (#713)
eigerco/polka-storage: rfcs/004-p2p-information-exchange.md

# RFC-004: P2P Information Exchange

Author: Aidan Sullivan - @aidan46
Date: 31/01/25

## Abstract

> This document covers the P2P [Peer ID][1] to [multi address][2] resolving information exchange and discusses possible improvements on the protocol.

Reviewer: ```suggestion
This document covers the P2P [Peer ID][1] to [multiaddress][2] resolving information exchange and discusses possible improvements on the protocol.
```

And "P2P [Peer ID][1] to [multi address][2] resolving information exchange" doesn't make a lot of sense to me — I don't get what you're trying to say.

I think the term causing confusion here is "information exchange", be specific

Author: Reworded this into something more specific, that bootstrap nodes are only aware of peers registered to them.

The documents outlines 2 different solutions; having bootstrap nodes exchange information between each other and connecting to all bootstrap nodes when resolving a [Peer ID][1] to a [multi address][2].

> The preferred solution would be to have bootstrap nodes gossip amongst themselves to ensure that the registration information is duplicated and recoverable.

Reviewer: ```suggestion
The preferred solution would be to have bootstrap nodes gossip amongst themselves to ensure that the registration information is replicated and recoverable.
```

## Problem Statement

The current design for the P2P information exchange is to have some storage providers bootstrap nodes and other register with these bootstrap nodes using the rendezvous protocol.
The problem with this design is that bootstrap nodes are not aware of each other and thus do not share registrations.
This isolates registered storage providers to the node they are registered to.

> When resolving the [Peer ID][1] to a [multi address][2] the connected bootstrap node can only share information of the peers that are registered to them.

Reviewer: Clients are the ones interested in resolving PeerIDs, not storage providers.

The problem statement is closer to "For clients looking for storage providers, they must first resolve the storage provider PeerID into a multiaddress, however, clients are only aware of collators, furthermore, as it currently stands, collators are only aware of the storage providers that registered directly with them"

Author: Reworded this so it includes the issue of clients resolving the mapping and only having partial information because the bootstrap nodes do not share registration information with each other.

## Proposals

### Bootstrap Node Collator Service

[Collators][3] maintain parachains by collecting parachain transactions from users and producing state transition proofs for relay chain validators.
Since collators are already well known within the network it makes sense for them to be bootstrap nodes an contain the P2P information.
The collators will act as [rendezvous][4] servers at which all the storage provider will register, providing information about their [Peer ID][1] and [multi address][2].

> The bootstrap node will be implemented as a service into the collator node.
> This ensures that all collators will run a bootstrap node.

Reviewer: Second sentence is redundant
```suggestion
The bootstrap node will be implemented as a service into the collator node.
```

Solving the bootstrap nodes not knowing the registration information that they each hold there are 2 options.

#### 1. Sharing information amongst bootstrap nodes

When the bootstrap nodes start up they will connect to each other over the P2P network.
After connecting with each other they will utilize libp2p's [gossipsub][5] protocol to relay registration information between themselves.
Information about new and updated registrations will be published over a topic and the bootstrap nodes will maintain an in-memory database that holds this information.

> The advantage of this solution is that all the registration information is duplicated.

Reviewer: ```suggestion
The advantage of this solution is that all the registration information is replicated.
```

This duplication makes sure that if a bootstrap node were to go down it can reconstruct the registration information from the other bootstrap nodes.
Furthermore, duplication ensures that when a client requests a [Peer ID][1] to [multi address][2] mapping it only has to request this information from a single bootstrap node.

> The disadvantage of this solution is the bootstrap nodes will have have a larger in-memory database, increasing the computational requirements for running a bootstrap node.

Reviewer 2: I don't think this is real disadvantage, it's a bunch of bytes.

However, since collators have [modern hardware requirements][6] the overhead of having an in-memory database with the registrations is minimal.

#### 2. Clients connecting to all bootstrap nodes

Bootstrap nodes split the registration information, holding only information of nodes that register with them.
Requiring the client, who wants to get a [Peer ID][1] to [multi address][2] mapping, to request this information from all bootstrap nodes.
This would require the bootstrap nodes to each have a database that holds their information in the event that they would go down.

The advantages of this that the implementation is simple, as it does not require any gossip between the bootstrap nodes.

The disadvantage of this solution is that resolving a [Peer ID][1] to a [multi address][2] will be slower as it requires connecting to all nodes in the worst-case scenario.

> Another disadvantage is that this requires bootstrap nodes to read and write to a database to ensure that information does not get lost in the event that they go down.

Reviewer: This still applies in the gossip scenario
